### PR TITLE
Make LivenessCheck Timeout Configurable

### PR DESCRIPTION
**Pull request by rajagopalanand**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by rajagopalanand** [comment]
Screenshot 2024-09-22 at 8 32 01 PM


Depending on the environment, the LivenessCheck could take longer than the original 100 ms like the graph above shows

**Review by yeya24 — COMMENTED** [review]
Let's make it configurable then?
There might be cases where 1s timeout couldn't work

**Comment by rapphil** [comment]
I think the liveness check should not be part of the client, because it is part of the business logic of the ruler, and has nothing to do with the connection which is the transport mechanism.

File: docs/guides/ruler-high-availability.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+linkTitle: "Ruler high availability"
+weight: 10
+slug: ruler-high-availability
+---
+
+This guide explains the concepts behind ruler high availability and when to use this feature
+
+## Background
+
+When rulers are deployed using shuffle sharding, each rule group is evaluated by a single ruler only. All the rulers in
```
Comment (yeya24) [inline docs/guides/ruler-high-availability.md:12]: Can we move the doc to its own PR?

File: pkg/ruler/ruler.go [hunk 2]
```diff
@@ -161,7 +159,8 @@ type Config struct {
 	EnableQueryStats      bool `yaml:"query_stats_enabled"`
 	DisableRuleGroupLabel bool `yaml:"disable_rule_group_label"`
 
-	EnableHAEvaluation bool `yaml:"enable_ha_evaluation"`
+	EnableHAEvaluation       bool          `yaml:"enable_ha_evaluation"`
+	EvalHAHealthCheckTimeout time.Duration `yaml:"eval_ha_health_check_timeout"`
```
Comment (yeya24) [inline pkg/ruler/ruler.go:163]: I don't like the flag name... But I don't have a good suggestion.

File: docs/guides/ruler-high-availability.md (hunk 1 above)
Comment (rajagopalanand) [reply docs/guides/ruler-high-availability.md:12]: Yes. Sorry about this. This was supposed to be in its own PR

File: pkg/ruler/ruler.go (hunk 2 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: Changed it to `LivenessCheckTimeout`

File: pkg/ruler/ruler.go [hunk 3]
```diff
@@ -238,6 +237,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.DisableRuleGroupLabel, "ruler.disable-rule-group-label", false, "Disable the rule_group label on exported metrics")
 
 	f.BoolVar(&cfg.EnableHAEvaluation, "ruler.enable-ha-evaluation", false, "Enable high availability")
+	f.DurationVar(&cfg.LivenessCheckTimeout, "ruler.liveness-check-timeout", 1*time.Second, "Liveness check timeout")
```
Comment (yeya24) [inline pkg/ruler/ruler.go:240]: Can we document the timeout itself. Like what is it for?

Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: Can you please take look? I made an update

**Review by yeya24 — APPROVED** [review]
Thanks!
